### PR TITLE
ARROW-6729: [C++] Prevent data copying in StlStringBuffer

### PR DESCRIPTION
**Pull request by st-pasha**

### CODE REVIEW

**Comment by wesm** [comment]
Can you open a JIRA issue for the changelog (https://github.com/apache/arrow/blob/master/docs/source/developers/contributing.rst#how-to-contribute-patches)? Might want to add a unit test to show that a `std::move`-d string is actually moved

**Comment by st-pasha** [comment]
Sure, I created a JIRA issue: https://issues.apache.org/jira/browse/ARROW-6729

**Review by wesm — APPROVED** [review]
+1. I added a check to the unit tests

File: cpp/src/arrow/buffer_test.cc [hunk 1]
```diff
@@ -173,11 +173,11 @@ TEST(TestMutableBuffer, Wrap) {
...
-  std::shared_ptr<Buffer> buffer;
-  {
-    std::string data_str = "input data";
-    buffer = Buffer::FromString(std::move(data_str));
-  }
+  std::string data_str = "input data";
+  std::shared_ptr<Buffer> buffer = Buffer::FromString(std::move(data_str));
+
+  // data_str is moved
+  ASSERT_EQ(0, data_str.size());
```
Comment (pitrou) [inline cpp/src/arrow/buffer_test.cc]: Instead, I would check that the pointer is the same.

**Comment by bkietz** [comment]
@wesm unfortunately that assertion is undefined behavior. The STL does not provide guarantees on the state of string after it has been moved (even though every impl I'm aware of would have size==0 after move)

File: cpp/src/arrow/buffer_test.cc (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/buffer_test.cc]: That will break in this case due to the short string optimization

Comment (pitrou) [reply cpp/src/arrow/buffer_test.cc]: Ahah.

Comment (bkietz) [reply cpp/src/arrow/buffer_test.cc]: But I think it's the right way to go: construct `expected` such that `expected.size() > sizeof(std::string)`, then store the data pointer before the move. It should be identical to `buffer->data()`

**Comment by wesm** [comment]
I see. I'll just remove the check then

**Pull request closed without merging**